### ISSUE TITLE
PEFT LoRA forwarding latency mismatch

### ISSUE THREAD

**Issue by leexcape** [huggingface/peft#2426]
### System Info

python 3.10.16
peft 0.14.1.dev0
transformers 4.49.0
torch = 2.6.0+cu126

### Who can help?

@BenjaminBossan 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [x] My own task or dataset (give details below)

### Reproduction

Hi there, I'm developing with PEFT LoRA to analyze the layer-wise inference latency. To do this, I placed a timestamp in `src/tuners/lora/layer.py`, line 712 and line 727 to record the forwarding latency of base layer and LoRA layer, respectively.

```python
else:
    time_base_layer_start = time.time() # timestamp here
    result = self.base_layer(x, *args, **kwargs)
    time_base_layer = time.time() - time_base_layer_start
    print('base layer inference time:', time_base_layer)
    torch_result_dtype = result.dtype
    for active_adapter in self.active_adapters:
        if active_adapter not in self.lora_A.keys():
            continue
        lora_A = self.lora_A[active_adapter]
        lora_B = self.lora_B[active_adapter]
        dropout = self.lora_dropout[active_adapter]
        scaling = self.scaling[active_adapter]
        x = self._cast_input_dtype(x, lora_A.weight.dtype)

        if not self.use_dora[active_adapter]:
            time_lora_layer_start = time.time() # timestamp here
            result = result + lora_B(lora_A(dropout(x))) * scaling
            time_lora_layer = time.time() - time_lora_layer_start
            print('lora layer inference time:', time_lora_layer)
```
In the results, the latency of base layer and lora layer doesn't show clear comparative relationship, which is shown in the results below:

> base layer inference time:  0.0001704692840576172
lora layer inference time:  0.00025725364685058594
base layer inference time:  5.9604644775390625e-05
lora layer inference time:  0.00013566017150878906

However, things are different in a stand-alone torch testing script. Below is a simple test script to analyze the forwarding latency of the `torch.nn.linear` of the same size with `self.base_layer()`, `lora_A` and `lora_B` in file layer.py. It worth mentioning that the `self.base_layer()`, `lora_A` and `lora_B` are also `torch.nn.linear` class. 

```python
import torch
import time
import numpy as np
import matplotlib.pyplot as plt
import datetime

# Set device (Modify here if you want to force CPU)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Define matrix sizes
a = 128  # Input batch size
b = 1024  # Large dimension
r = 64  # Smaller rank for LoRA

# Initialize random tensors on the selected device
x = torch.randn(a, b, device=device)
M = torch.randn(b, b, device=device)  # Full-rank matrix
lora_A = torch.randn(b, r, device=device)  # LoRA A (Low-rank)
lora_B = torch.randn(r, b, device=device)  # LoRA B (Low-rank)
dropout = torch.nn.Dropout(p=0.1).to(device)  # Dropout layer
scaling = 0.1
y = torch.randn(a, b, device=device)  # Bias term for 0-LoRA

# Define different inference functions
def base_layer_inference():
    return torch.matmul(x, M)

def lora_inference_0():
    return torch.matmul(torch.matmul(dropout(x), lora_A), lora_B) * scaling + y

def lora_inference_1():
    return torch.matmul(torch.matmul(dropout(x), lora_A), lora_B) * scaling

def lora_inference_2():
    return torch.matmul(torch.matmul(dropout(x), lora_A), lora_B)

def lora_inference_3():
    return torch.matmul(x, lora_A).matmul(lora_B)

# Measure latency for each function
def measure_latency(func, iterations=1000):
    latencies = []
    for _ in range(iterations):
        start = time.time()
        _ = func()
        torch.cuda.synchronize() if device.type == "cuda" else None  # Ensure accurate timing
        latencies.append((time.time() - start) * 1000)  # Convert to ms
    return np.array(latencies)

# Run latency measurements
iterations = 1000
lat_base = measure_latency(base_layer_inference, iterations)
lat_0 = measure_latency(lora_inference_0, iterations)
lat_1 = measure_latency(lora_inference_1, iterations)
lat_2 = measure_latency(lora_inference_2, iterations)
lat_3 = measure_latency(lora_inference_3, iterations)

# Plot the results
plt.figure(figsize=(12, 6))
plt.plot(lat_base, label=f"Base Layer (Mean: {lat_base.mean():.3f} ms, Std: {lat_base.std():.3f})", alpha=0.8)
plt.plot(lat_0, label=f"0-LoRA Layer (Mean: {lat_0.mean():.3f} ms, Std: {lat_0.std():.3f})", alpha=0.8)
plt.plot(lat_1, label=f"1-LoRA Layer (Mean: {lat_1.mean():.3f} ms, Std: {lat_1.std():.3f})", alpha=0.8)
plt.plot(lat_2, label=f"2-LoRA Layer (Mean: {lat_2.mean():.3f} ms, Std: {lat_2.std():.3f})", alpha=0.8)
plt.plot(lat_3, label=f"3-LoRA Layer (Mean: {lat_3.mean():.3f} ms, Std: {lat_3.std():.3f})", alpha=0.8)

# Customize plot
plt.title("Inference Latency per Iteration")
plt.xlabel("Iteration")
plt.ylabel("Latency (ms)")
plt.legend()
plt.grid(True)

# Save figure as a high-quality JPG file
timestamp = datetime.now().strftime("%Y%m%d_%H_%M_%S_%f")[:-3]
filename = "results/latency_test_1_" + timestamp + ".jpg"
plt.savefig(filename, dpi=300, bbox_inches='tight')

print(f"Plot saved as {filename}")

```

The results showed that the latency of base layer is greater than that of lora layers, which makes more sense because the lora consists less parameters. 

![Image](https://github.com/user-attachments/assets/043146d4-e78c-4f18-8e05-67c9e81d6297)

In the 1000 runs shown in this figure, all the lora operations are faster than the base model, and the more operations there are in the lora forwarding, the higher latencies there are. I have no clue the PEFT lora and the stand-alone test script show such difference on the computational capacity. Any help or suggestions would be greatly appreciated!

### Expected behavior

The forwarding latency of `lora_A` and `lora_B` should be smaller than that of the `base_layer`, which makes more sense theoretically and has been validated by the stand-alone torch testing script.

**Comment by BenjaminBossan** [comment]
Thank you for investigating the runtime performance of LoRA. Regarding your results, I'm not sure if this is an apples to apples comparison, especially since you didn't show how exactly you measured the original results. Some potential issues that come to mind:

1. If you measured the original results during training, then you have to account for the whole autograd machinery being active (note that only LoRA collects gradients, not the base model). In your second script, it is completely absent.
2. Did you match the sizes of the parameters/tensors?
3. I'm uncertain if the timing measurements are correct as is, since they may involve CUDA calls and the values are really small.

**Comment by BenjaminBossan** [comment]
@leexcape did you have time to look into this further?
